### PR TITLE
[ENG-5168] User addon credentials

### PR DESCRIPTION
**Pull request by futa-ikeda**
-   Ticket: [ENG-5168]
-   Feature flag: n/a

## Purpose
- Allow users to add new addon accounts in their user settings page

## Summary of Changes
- Make node argument optional for Provider class
- Update user-addons-manager to handle auth workflows
- Update external-service models
- Update user-settings addon page to allow for auth workflows

## Screenshot(s)
- Desktop (adding new Amazon S3 account)
image

- Mobile (adding new Dataverse account)
image

## Side Effects
- Hopefully none and the auth workflow for project addons works great still

## QA Notes

[ENG-5168]: https://openscience.atlassian.net/browse/ENG-5168?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/settings/addons/template.hbs [hunk 1]
```diff
@@ -5,57 +5,107 @@
     {{#if manager.currentListIsLoading}}
         <LoadingIndicator @dark={{true}} />
     {{else}}
-        <div local-class='addons-list-wrapper {{if this.isMobile 'mobile'}}'>
-            <div
-                data-analytics-scope='Addon List Filter'
-                local-class='filter-wrapper {{if this.isMobile 'mobile'}}'
+        {{#if manager.selectedProvider}}
```
Comment (futa-ikeda) [inline app/settings/addons/template.hbs:8]: A lot of the look/feel of this is subject to change as we decide on mocks

File: app/packages/addons-service/provider.ts [hunk 2]
```diff
@@ -83,8 +87,8 @@ export default class Provider {
         return Boolean(this.configuredAddon);
     }
 
-    constructor(provider: any, currentUser: CurrentUserService, node: NodeModel) {
-        setOwner(this, getOwner(node));
+    constructor(provider: any, currentUser: CurrentUserService, node?: NodeModel) {
+        setOwner(this, getOwner(provider));
```
Comment (brianjgeiger) [inline app/packages/addons-service/provider.ts:91]: Did typescript help you catch this, or did you get an error of some sort and had to figure out what it was from?

File: app/packages/addons-service/provider.ts [hunk 3]
```diff
@@ -197,7 +203,7 @@ export default class Provider {
     }
 
     async userAddonAccounts() {
-        return await this.userReference.authorizedStorageAccounts;
+        return await this.userReference?.authorizedStorageAccounts;
```
Comment (brianjgeiger) [inline app/packages/addons-service/provider.ts:206]: Is there a way we wouldn't have a userReference?

File: lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts [hunk 4]
```diff
@@ -137,13 +177,46 @@ export default class UserAddonManagerComponent extends Component<Args> {
         }
     }
 
+    @task
+    @waitFor
+    async getAuthorizedCitationAccounts() {
+        const { userReference } = this;
```
Comment (brianjgeiger) [inline lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts:183]: I see you are embracing vaguely modern JS syntax. Very nice.

File: lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts [hunk 5]
```diff
@@ -137,13 +177,46 @@ export default class UserAddonManagerComponent extends Component<Args> {
...
+        const mappedObject = this.filterTypeMapper[FilterTypes.CITATION_MANAGER];
+        if (userReference) {
```
Comment (brianjgeiger) [inline lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts:185]: Especially for the user-addons-manager, which would only be on the user settings page, shouldn't we always have a userReference here? Or is it a timing issue while the API request is being made?

**Review by brianjgeiger — APPROVED** [review]
A couple of questions (or maybe one question twice), a couple of comments, but looks good.

File: app/packages/addons-service/provider.ts (hunk 2 above)
Comment (futa-ikeda) [reply app/packages/addons-service/provider.ts:91]: I got an error when I tried to do `getOwner` on the node, now that it's an optional parameter (not used on user addons page)

File: app/packages/addons-service/provider.ts (hunk 3 above)
Comment (futa-ikeda) [reply app/packages/addons-service/provider.ts:206]: Ahh, good point. I guess in the very off-chance that the request to get the userReference from GV fails, maybe? I don't think this change is needed though, and is a leftover from some earlier type changes that I was trying out?

File: lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts (hunk 5 above)
Comment (futa-ikeda) [reply lib/osf-components/addon/components/addons-service/user-addons-manager/component.ts:185]: There _should_ be a userReference, as it is the first thing done when we initialize this component, so I don't think there should be a timing issue. I've defined userReference as an optional property on this (`userReference?: UserReferenceModel`), just to be safe, and these `if` statements are a consequence of that typing. I think I'm leaning towards doing `userReference!:` though, as this component doesn't make much sense without it, as you said

**Pull request merged**